src/bson_mapper/bson_archiver.hpp

```diff
@@ -0,0 +1,991 @@
+            auto newNode = search();
+            if (newNode.type() == bsoncxx::type::k_document) {
+                _embeddedBsonDocStack.push(newNode.get_document().value);
+                _nodeTypeStack.push(InputNodeType::InEmbeddedObject);
+            } else if (newNode.type() == bsoncxx::type::k_array) {
+                _embeddedBsonArrayStack.push(newNode.get_array().value);
+                _embeddedBsonArrayIteratorStack.push(_embeddedBsonArrayStack.top().begin());
+                _nodeTypeStack.push(InputNodeType::InEmbeddedArray);
+            } else {
+                throw cereal::Exception("Node requested is neither document nor array.\n");
+            }
+        }
+    }
+
+    /**
+     * Finishes the most recently started node by popping relevant stacks
+     * and, if necessary, iterating to the next root BSON document.
+     */
+    void finishNode() {
+        // If we're in an embedded object or array, pop it from its respective
+        // stack(s).
+        if (_nodeTypeStack.top() == InputNodeType::InEmbeddedObject) {
+            _embeddedBsonDocStack.pop();
+        } else if (_nodeTypeStack.top() == InputNodeType::InEmbeddedArray) {
+            _embeddedBsonArrayStack.pop();
+            _embeddedBsonArrayIteratorStack.pop();
+        }
+
+        // Pop the node type from the stack.
+        _nodeTypeStack.pop();
+
+        if (!_nodeTypeStack.empty() && _nodeTypeStack.top() == InputNodeType::InRootElement) {
+            _nodeTypeStack.pop();
+        }
+    }
+
+    /**
+     * Sets the name for the next node created with startNode
+     *
+     * @param name
+     *  The name of the next node
+     */
+    void setNextName(const char* name) {
+        _nextName = name;
+    }
+
+   private:
+    /**
+     * Throws an exception if the type of v is not the specified type t.
+     */
+    inline void assert_type(const bsoncxx::types::value& v, bsoncxx::type t) {
+        if (v.type() != t) {
+            throw cereal::Exception("Type mismatch when loading values.\n");
+        }
+    }
+
+   public:
+/**
+ * Loads a BSON value from the current node into a bsoncxx::types variable.
+ *
+ * @param val
+ *    The bsoncxx::types typed variable into which the BSON value will be loaded.
+ */
+#define BSON_LOAD_VALUE_FUNC(btype)                     \
+    void loadValue(bsoncxx::types::b_##btype& val) {    \
+        auto bsonVal = search();                        \
+        assert_type(bsonVal, bsoncxx::type::k_##btype); \
+        val = bsonVal.get_##btype();                    \
+    }
+
+    // Invokes the macro for all non-deprecated, non-internal
+    // BSON types supported by bsoncxx
+    BSON_LOAD_VALUE_FUNC(double)
+    BSON_LOAD_VALUE_FUNC(utf8)
+    BSON_LOAD_VALUE_FUNC(document)
+    BSON_LOAD_VALUE_FUNC(array)
+    BSON_LOAD_VALUE_FUNC(binary)
+    BSON_LOAD_VALUE_FUNC(oid)
+    BSON_LOAD_VALUE_FUNC(bool)
+    BSON_LOAD_VALUE_FUNC(date)
+    BSON_LOAD_VALUE_FUNC(int32)
+    BSON_LOAD_VALUE_FUNC(int64)
+
```

Comment:
Technically you have to do more to be really hygenic (pragma push before, then undef before defining yours, then undef, then pragma pop on the way out).

Probably better to disambiguate these (MONGOCXX_ODM_BSON_LOAD_VALUE_FUNC, or whatever other good prefix we're picking for this project) and not worry about it